system-addon/lib/AutoMigrateFeed.jsm

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+const {actionTypes: at, actionCreators: ac} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+
+XPCOMUtils.defineLazyModuleGetter(this, "AutoMigrate",
+  "resource:///modules/AutoMigrate.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "Services",
+  "resource://gre/modules/Services.jsm");
+
+this.AutoMigrateFeed = class AutoMigrateFeed {
+  onAction(action) {
+    switch (action.type) {
+      case at.AUTOMIGRATE_MIGRATE_DONE:
+        AutoMigrate.keepAutoMigration();
+        this.store.dispatch(ac.BroadcastToContent({type: at.AUTOMIGRATE_HIDE}));
+        break;
+      case at.AUTOMIGRATE_UNDO_MIGRATION:
+        AutoMigrate.undoAutoMigration(this.window);
```

Comment:
I think it might be better to use `action._target.browser` (that is, from the `AUTOMIGRATE_UNDO_MIGRATION` action), since there could be multiple windows open at once